File: python/mirage/commands/builtin/generic/crossmount/relay/sort.py
from functools import partial
from typing import cast

from mirage.commands.builtin.generic.crossmount.types import CrossResult
from mirage.commands.builtin.generic.crossmount.utils import (_relay_write,
                                                              flat_scopes,
                                                              relay)
from mirage.commands.builtin.generic.sort import sort as generic_sort
from mirage.commands.spec.types import FlagValue
from mirage.io.types import ByteSource
from mirage.runtime.types import DispatchFn
from mirage.types import FileType, PathSpec


async def run_sort(scopes: list[PathSpec], flags: dict[str, FlagValue],
                   dispatch: DispatchFn,
                   stdin: ByteSource | None) -> CrossResult:
    """Sort independent inputs through the same generic on every mount.

    Args:
        scopes (list[PathSpec]): Expanded operands in command-line order.
        flags (dict[str, FlagValue]): Parsed options.
        dispatch (DispatchFn): Workspace operation dispatcher.
        stdin (ByteSource | None): Shared standard input cursor.
    """

    async def read_file(path: PathSpec) -> bytes:
        info = await relay(dispatch, "stat", path)
        if info.type is FileType.DIRECTORY:
            raise IsADirectoryError(path.virtual)
        return cast(bytes, await relay(dispatch, "read", path))

    return await generic_sort(flat_scopes(scopes),
                              read_bytes=read_file,
                              write_bytes=partial(_relay_write, dispatch),
                              stdin=stdin,
                              flags=flags)

File: typescript/packages/core/src/commands/builtin/generic/crossmount/relay/sort.ts
import { IOResult, type ByteSource } from '../../../../../io/types.ts'
import { type PathSpec } from '../../../../../types.ts'
import type { FlagValue } from '../../../../spec/types.ts'
import { sortGeneric } from '../../sort.ts'
import type { CrossResult, DispatchFn } from '../types.ts'
import { crossOpts, flatten, fileStreamOp } from '../utils.ts'

export async function runSort(
  scopes: PathSpec[],
  bag: Record<string, FlagValue>,
  dispatch: DispatchFn,
  stdin: ByteSource | null,
): Promise<CrossResult> {
  const reads = new IOResult()
  const write = async (path: PathSpec, data: Uint8Array): Promise<void> => {
    await dispatch('write', path, [data])
    // A replacement must cache the new bytes, not its earlier input.
    Reflect.deleteProperty(reads.reads, path.virtual)
  }
  const [body, io] = (await sortGeneric(
    flatten(scopes),
    { ...crossOpts(bag), stdin },
    fileStreamOp(dispatch, reads),
    write,
  )) ?? [null, new IOResult()]
  return [body, await reads.merge(io)]
}
